### ISSUE TITLE
Custom Save Location for option 'addable-box'

### ISSUE THREAD

**Issue by nicolaelitvac** [ThemeFuse/Unyson#2883]
I have an option like this:
```php
'ringGallery' => array(
    'type'  => 'addable-box',
    'value' => array(),
    'attr'  => array( 'class' => 'custom-class' ),
    'label' => __('Ring Gallery', 'insignety-rings'),
    'desc'  => __('Ring Gallery images.', 'insignety-rings'),
    //'help'  => __('Help tip', '{domain}'),
    'box-options' => array(
        'thumb' => array( 'type' => 'text' ),
        'full' => array( 'type' => 'text' )
    ),
    'template' => 'Image', // box title
    'box-controls' => array( // buttons next to (x) remove box button
        //'control-id' => '<small class="dashicons dashicons-smiley"></small>',
    ),
    'limit' => 0, // limit the number of boxes that can be added
    'add-button-text' => __('Add', 'insignety-rings'),
    'sortable' => false,
    'fw-storage' => array(
        'type' => 'post-meta',
        'post-meta' => 'ringGallery'
    )
)
```

I'd like that to be saved as serialised array in post-meta "ringGallery".

Is that possible currently? Or are you planning on adding such feature?

**Comment by ViorelEremia** [comment]
why do you want to do this?

**Comment by nicolaelitvac** [comment]
I'm creating posts with API.
My JSON contains:
```json
"ringGallery": [
  {
    "full": "http://insignetydev.wpengine.com/wp-content/uploads/2017/09/ring-demo.jpg",
    "thumb": "http://insignetydev.wpengine.com/wp-content/uploads/2017/09/ring-demo-150x150.jpg"
  },
  {
    "full": "http://insignetydev.wpengine.com/wp-content/uploads/2017/09/ring-demo.jpg",
    "thumb": "http://insignetydev.wpengine.com/wp-content/uploads/2017/09/ring-demo-150x150.jpg"
  }
]
```

If you'd enable saving (sync) with a post-meta for this option type, I'll be able to populate those fields with the JSON containing the above part when making the POST request to REST API.

**Comment by nicolaelitvac** [comment]
I think something like the filters that where created for https://github.com/ThemeFuse/Unyson/issues/1440 might save the day.

**Comment by nicolaelitvac** [comment]
Or maybe I should use another function for saving that kind of meta on post creation, currently I use `fw_set_db_post_option()` ?

**Comment by ViorelEremia** [comment]
Ok I tried that in rest api wp: 
`function try_register_rest_field( $object, $field_name, $request ) {
	return fw_get_db_post_option( get_the_ID(), 'ringGallery', array() );
}

function try_rest_api_init() {
	register_rest_field(
		'post',
		'ringGallery',
		array(
			'get_callback'    => 'try_register_rest_field',
			'update_callback' => null,
			'schema'          => null
		)
	);
}

add_action( 'rest_api_init', 'try_rest_api_init' );`
then i acccess my post via wp-json everythin is fine exemple: http://localhost/wp/wp-json/wp/v2/posts/mypostid I see key here.
Then you save optin via ajax then you save ordinary but then you send in browser you encode this extracted value as : `wp_send_json( array( 'ringGallery' => fw_get_db_post_option( get_the_ID(), 'ringGallery', array() ) ) );`
in short I do not see the problem.

**Comment by nicolaelitvac** [comment]
The issue is not with getting the data, but saving/putting it.
Add the `update_callback` in `try_rest_api_init()`.
Then try to POST this json:
```json
[
  {
    "ringGallery": [
      {
        "full": "http://insignetydev.wpengine.com/wp-content/uploads/2017/09/ring-demo.jpg",
        "thumb": "http://insignetydev.wpengine.com/wp-content/uploads/2017/09/ring-demo-160x160.jpg"
      },
      {
        "full": "http://insignetydev.wpengine.com/wp-content/uploads/2017/09/ring-demo.jpg",
        "thumb": "http://insignetydev.wpengine.com/wp-content/uploads/2017/09/ring-demo-160x160.jpg"
      }
    ],
    "content": "Voluptate dolore consectetur ...",
    "title": "Ring-1",
    "status": "publish"
  }
]
```

**Comment by ViorelEremia** [comment]
Please provide your code I can't reproduce it in my end.

**Comment by nicolaelitvac** [comment]
Hi Viorel,
I've created this demo theme from Unyson starter to showcase the issue. I've updated `functions.php` and created `ring.php` in framework-customization for meta-boxes.
Theme: https://drive.google.com/open?id=0B6GU_sfH8LqoVE0tQ2lWMkliRFk
After theme was activated, access any url of the website with query `?sync=true`. Ex: `http://domain.dev/wp-admin/?sync=true` you should see the an array with Created/Updated post IDs.
This is the demo data that's imported.
JSON: http://myjson.com/1fr8z5

After the import, when I try to edit any post (CPT 'ring') then I get this error:
`Recoverable fatal error: Object of class stdClass could not be converted to string in .../wp-content/plugins/unyson/framework/includes/option-types/simple.php on line 235`.

**Comment by nicolaelitvac** [comment]
I've updated a bit `addable-box` and it works now.

Meta field declaration:
```php
'ringGallery' => array(
    'type'  => 'addable-box',
    'value' => array(),
    'attr'  => array( 'class' => 'custom-class' ),
    'label' => __('Ring Gallery', 'insignety-rings'),
    'desc'  => __('Ring Gallery images.', 'insignety-rings'),
    //'help'  => __('Help tip', '{domain}'),
    'box-options' => array(
        'thumb' => array(
        	'type' => 'text'
        ),
        'full' => array( 'type' => 'text' )
    ),
    'template' => 'Image', // box title
    'limit' => 0, // limit the number of boxes that can be added
    //'width' => 'full',
    'add-button-text' => __('Add', 'insignety-rings'),
    'sortable' => false,
    'fw-storage' => array(
        'type' => 'post-meta',
        'post-meta' => 'ringGallery'
    )
)
``` 

Updated line 76 in `class-fw-option-type-addable-box.php` with:
```php
if (empty($data['value']) || !is_array( maybe_unserialize( $data['value']) ) ) {
```

Added a line after line 31 in `addable-box/view.php' with:
```php
$data['value'] = maybe_unserialize( $data['value'] );
```

**Comment by nicolaelitvac** [comment]
But in order for the 'meta_query' to work this has to be saved in the WP's default location somehow.

**Comment by ViorelEremia** [comment]
@nicolaelitvac sorry I was very busy, but if is important to you create a pull request with a detailed description of what you did, this will speed up the process of solving the issue.
